### PR TITLE
feat(loans): accrue monthly interest so payments reduce principal only (#1804)

### PR DESCRIPTION
**Pull request by thewillwilson**
Refs #1804 — implements the auto-split half. The schedule/visibility half is open in #1887 (see *Relationship to #1887* below).

## Problem

Linking a mortgage payment as a transfer into a Loan account reduces the principal by the **full payment amount** — `Balance::BaseCalculator#derive_non_cash_balance` applies loan transactions straight to the non-cash balance. For a mortgage, most of an early payment services interest, so the tracked balance falls far faster than the real one and net worth drifts steadily optimistic.

The only workaround today is manually splitting the payment against the lender's amortization table every month. Because the principal/interest ratio changes every month, that is a permanent monthly chore — exactly what #1804 asks to remove.

## Approach

Rather than rewriting the user's payment transaction — which would fight bank sync, the split machinery and transfer matching — post the *other half* of the ledger: a monthly "Interest charged" entry on the loan account that increases the balance, so the payment nets down to principal only.

```
01 Jul  Interest charged        +833.33
01 Jul  Payment from Checking   -849.00
------------------------------------------
Principal reduction              -15.67
```

- **Charged on the balance actually outstanding on the accrual date**, not a schedule projected from origination. Overpayments, missed payments and irregular schedules self-correct on the next accrual.
- **Charging starts at an explicit `interest_accrual_start_date`**, never at the account's opening anchor. The two are not the same date: an opening anchor records when a balance was first *known*, and `AccountableResource#create` backdates it two years when the user supplies none. Accruing from the anchor would invent years of debt the borrower never incurred.
- **Opt-in per loan, manual accounts only.** A linked account is anchored to the principal its provider reports, so a replay from origination has no reliable starting point. Enabling accrual and later linking the account purges the accruals rather than leaving them to inflate a provider-anchored balance.

### Variable and adjustable rates (effective-dated)

Rate resets are modelled as effective-dated `rate_changes` rather than a single mutable rate. The base `interest_rate` applies from origination and each `Loan::RateChange (effective_date, rate)` resets the rate from its effective date forward, so `Loan::InterestAccrual` prices each period at the rate the lender actually charged then — an ARM reset reprices only periods on or after it, not the whole history.

Editing the base `interest_rate` itself still reprices everything, which is the right behavior for correcting a mis-entered origination rate. (This replaces the earlier "fixed-rate in practice" limitation flagged in @jjmata's review — the retroactive-repricing concern is resolved by pricing each period at its effective-dated rate.) Rate changes are managed from a small add/remove form on unlinked loans; no API surface.

### Why the entries are `excluded`, and why that isn't sufficient on its own

`Balance::SyncCache` does not filter on `excluded`, so accrual entries still move the balance; `IncomeStatement` does filter on it, so the interest is not double-counted against the payment already booked as a `loan_payment` expense on the funding account.

But `excluded` is **not** a general "hide this" flag in this codebase — it is honored by `IncomeStatement` and the account activity tab, and ignored by the family-wide transactions page, the rules engine and CSV export. Left alone, accruals would inflate the transactions-page expense total, get renamed or recategorized by any family rule matching "interest" on every sync, and leak into exports as if they were the user's own transactions.

This PR therefore adds `Entry.excluding_system_generated` alongside the existing `Entry.excluding_split_parents`, and applies it in `Transaction::Search`, `Rule::Registry::TransactionResource` and both `Family::DataExporter` paths. Generated entries are also fully re-asserted on every sync — `date`, `amount`, `currency`, `name` and `excluded` — since they are derived output, and an un-excluded accrual would otherwise double-count permanently with no path back.

One drive-by fix: `Entry#linked?` is only `external_id.present?`, so the transactions header rendered a "Linked with " badge (empty provider name) for any entry carrying an external id on a provider-less account. Now gated on the account actually having a provider.

## Migrations

- `20260729000000_add_interest_accrual_to_loans` (`Migration[7.2]`) — adds `loans.accrue_interest` (boolean, `default: false`, `null: false`), `loans.interest_accrual_day` (integer, nullable) and `loans.interest_accrual_start_date` (date, nullable), plus a `loans_interest_accrual_day_check` constraint bounding the day to 1..31 mirroring the AR validation.
- `20260730000000_create_loan_rate_changes` (`Migration[7.2]`) — adds the `loan_rate_changes` table (`loan_id` FK, `effective_date`, `rate` decimal(10,3)), a unique index on `(loan_id, effective_date)`, and a `rate >= 0` check constraint mirroring the AR validation.

`db/schema.rb` is edited to exactly the lines these migrations produce (verified against a real `bin/rails db:migrate` run on this branch). It is deliberately *not* a wholesale regenerated dump: the installed Rails 8.1 dumper rewrites `Schema[7.2]` → `[8.1]`, `plpgsql` → `pg_catalog.plpgsql` and reorders columns across dozens of unrelated tables, producing ~1,600 lines of churn of the sort e75f2a0c was pushed to undo. Happy to take whichever form maintainers prefer.

## Testing

```
$ bin/rails test
6053 runs, 24139 assertions, 0 failures, 0 errors, 30 skips

$ bin/rubocop
2164 files inspected, no offenses detected

$ bundle exec erb_lint app/views/loans/_form.html.erb \
                       app/views/loans/tabs/_overview.html.erb \
                       app/views/transactions/_header.html.erb
Linting 3 files with 16 linters...
No errors were found in ERB files

$ bin/brakeman --no-pager
Errors: 0
Security Warnings: 0
No warnings found
```

(`erb_lint --lint-all` reports 9 offenses on `main`, all in files this PR does not touch — none in the three above.)

New coverage:

- `test/models/loan/interest_accrual_test.rb` — the interest/principal split and its effect on the materialized balance; charging only from the configured start date (with a regression case for a two-year-backdated opening anchor); skipping valuation dates; honoring a payment's custom exchange rate; repairing a user-edited accrual; idempotent regeneration; re-pricing on a rate change; purge on opt-out and on linking; month-end clamping; payoff; and a behavioural assertion that accrued interest does **not** move the `IncomeStatement` expense total.
- `test/models/loan_test.rb` — `accrues_interest?` gating, validation of accrual inputs, resync-on-change.
- `test/models/account/syncer_test.rb` — accrual runs before materialization, drops the incremental window when accruals changed, keeps it when they didn't, degrades rather than failing the sync on error, and is skipped entirely for non-loan accounts.
- `test/controllers/loans_controller_test.rb` — the new permitted params persist through `AccountableResource#update` (on and off), enabling without a rate or start date is rejected, and the toggle renders only for unlinked loans.

## Relationship to #1887

#1887 implements the *schedule* half of #1804 and overlaps this branch on `app/models/loan.rb`, `app/views/loans/tabs/_overview.html.erb`, `config/locales/views/loans/en.yml` and `test/models/loan_test.rb`.

The two are complementary rather than duplicative — that PR projects a schedule for display, this one posts real ledger entries — and the overlaps are additive in each file (both add summary cards to the same `grid-cols-3` container). Both currently merge cleanly against `main`, so whichever lands second just needs a rebase. **Happy to rebase on top of #1887 if maintainers want that one first** — no objection to it going ahead, and I'd rather not step on @Rene0422's work.

Worth noting that @jjmata's point 2 on #1887 — that deriving a schedule start from `account.first_valuation&.date` surfaces dates years in the past for an existing mortgage — is the same modelling issue this PR ran into, and is why charging here begins at an explicit `interest_accrual_start_date` rather than at the opening anchor. That feedback also informed the choices here to use `BigDecimal` throughout rather than `Float`, and `DS::*` components rather than raw markup.

## i18n

New keys added to `config/locales/views/loans/en.yml` only; other locales fall back per `config.i18n.fallbacks` and `config/i18n-tasks.yml`, consistent with #2794, #2727 and #2636.

## Notes

- No new design-system primitives. The form reuses `StyledFormBuilder#toggle`, `shared/ruler` and `DS::Alert`; the Overview tab reuses `summary_card` and `DS::Link`.
- No new API schema or fields, but `Api::V1::TransactionsController` does change behaviour on an existing endpoint: `excluding_system_generated` filtering on index/show/update/destroy plus a new `422 "Source is reserved"` branch on create. Per the repo's API-consistency convention this is covered by Minitest behavioural tests in `test/controllers/api/v1/transactions_controller_test.rb` and a docs-only rswag example in `spec/requests/api/v1/transactions_spec.rb` so `docs/api/openapi.yaml` reflects the reserved-source validation error. (Thanks @jjmata — corrects the earlier "no API surface" note.)
- Accrual entries are identified by **period** (`2026-02`) rather than by date, so changing the statement day or start date re-dates existing rows in place rather than destroying and recreating them — on a 27-year loan, ~340 updates instead of ~340 `dependent: :destroy` cascades plus ~340 inserts. A test pins the "at most one accrual per calendar month" invariant that this identity relies on.
- Genuine purges (opting out, or linking the account) still go through `destroy_all` and cost ~13 queries per entry via `Entry`'s `dependent: :destroy` fan-out — roughly 1s for a 27-year loan. Left as-is deliberately: these are once-ever deliberate actions, and a hand-rolled bulk delete would have to replicate six cascade paths including Active Storage blob purging, which silently rots the moment anyone adds a new `dependent: :destroy`. Happy to revisit if reviewers disagree.

## Summary by CodeRabbit

* **New Features**
  * Added loan interest accrual settings (enable toggle, start date, accrual day) for eligible (unlinked) loans, plus “interest charged” display on the loan overview.
  * Introduced automated monthly interest accrual generation and resync behavior when accrual inputs change.
  * Preserved accrual settings during data import.
* **Bug Fixes**
  * Excluded system-generated accrual entries from transactions UI/search/export and the transactions API, including id lookup, show/update/destroy, and reserved idempotency handling.
* **Tests**
  * Added/expanded coverage for accrual syncing, validations, UI visibility, idempotency, export/search/API filtering, and accrual math edge cases.

### CODE REVIEW

**Comment by thewillwilson** [comment]
Heads-up before anyone spends review time here: I opened this without having seen **#2840**, which covers the same ground and was opened a day earlier. My search was for "mortgage" / "amortization" and didn't match its title — that's my mistake, not a deliberate parallel effort. Had I found it I'd have commented there rather than opening this.

Both implement the auto-split half of #1804, by different mechanisms:

- **#2840** splits the cash-side payment into a principal transfer plus an interest expense categorized as *Loan Interest*.
- **This PR** leaves the payment untouched and posts a monthly "Interest charged" entry on the loan account, so the payment nets down to principal.

**Where #2840 is better:** it's what #1804 literally asked for, and it surfaces interest as a real category in cash flow and budgets. This PR deliberately keeps the charge out of `IncomeStatement` — the payment is already booked as a `loan_payment` expense, so counting the interest again would double it — which means interest only appears on the loan's overview card. That's strictly less useful for anyone wanting to see the true cost of their debt in reports.

**Where this one differs:** it never modifies the user's payment transaction, so it doesn't interact with bank sync, the split machinery, or transfer matching; each charge is re-derived from the balance actually outstanding on the accrual date rather than computed once at match time; and because it reads the configured rate each period, it works for variable and adjustable rates rather than fixed only.

**I'm happy to close this in favour of #2840** — it came first and already has @jjmata's review. Just say the word. If the accrual model is interesting as a *complement* rather than a competitor (variable-rate loans, or provider-synced payments that shouldn't be rewritten), I'd gladly rework it to sit alongside that PR instead.

---

One change here is independent of either design, and I'd suggest pulling it out regardless of what happens to the rest:

`Entry#linked?` is only `external_id.present?`, so `app/views/transactions/_header.html.erb` renders a "Linked with " badge — with an empty provider name — for any entry carrying an external id on an account that has no provider (`Account#provider_name` returns `nil`). It's a one-line gate on `provider_name.present?` and has nothing to do with loans. Happy to open it as its own PR.

File: app/models/entry.rb [hunk 1]
```diff
@@ -79,6 +79,20 @@ class Entry < ApplicationRecord
...
+  # balance correct, and are re-asserted from scratch on every sync.
+  #
+  # `excluded` alone is not enough to keep them out of the way — it is honored by
+  # IncomeStatement and the account activity tab, but not by the family-wide
+  # transactions page, the rules engine, or CSV export. Anything that presents
+  # entries as the user's own transactions should apply this scope.
+  SYSTEM_GENERATED_SOURCES = %w[loan_interest_accrual].freeze
+
+  scope :excluding_system_generated, -> {
+    where("entries.source IS NULL OR entries.source NOT IN (?)", SYSTEM_GENERATED_SOURCES)
```
[the remark this answers is not in the document]
Comment (thewillwilson) [reply app/models/entry.rb:103]: Confirmed and fixed in e8bd8f88 — good catch. I applied `excluding_system_generated` to the web search, rules and export paths and missed the API entirely.

`Api::V1::TransactionsController#index` now merges the scope. Added a test asserting an accrual entry does not appear in `GET /api/v1/transactions`.

File: app/models/family/data_exporter.rb [hunk 2]
```diff
@@ -464,11 +464,16 @@ def generate_ndjson
...
+             .merge(Entry.excluding_split_parents)
+             .merge(Entry.excluding_system_generated)
     end
 
     def ndjson_exportable_transactions
-      @family.transactions.joins(:entry).where(entries: { parent_entry_id: nil })
+      @family.transactions
+             .joins(:entry)
+             .where(entries: { parent_entry_id: nil })
+             .merge(Entry.excluding_system_generated)
```
[the remark this answers is not in the document]
Comment (thewillwilson) [reply app/models/family/data_exporter.rb:496]: Confirmed and fixed in e8bd8f88.

The accrual settings now round-trip through `Family::DataImporter#import_accounts`, so a restored loan knows to rebuild the entries the export deliberately omits. Test added covering the restore.

Two notes for the reviewer:

- `interest_rate` had to come along too, not just the accrual fields — `Loan` validates its presence when accrual is enabled, so restoring the toggle without it would fail the whole import rather than just the loan.
- `rate_type` and `term_months` are still dropped on restore. That gap predates this PR (`safe_accountable_attrs` was `%w[subtype locked_attributes]` before it), so I have left it alone rather than widening this diff. Happy to fix it here or separately if maintainers prefer.

File: app/models/account/syncer.rb [hunk 3]
```diff
@@ -17,6 +18,25 @@ def perform_post_sync
...
+    # the window, which would otherwise be seeded from a now-stale persisted
+    # balance. Failures are swallowed — an accrual problem should degrade the loan
+    # to its pre-existing behaviour, not fail the whole account sync.
+    def accrue_loan_interest
+      return false unless account.loan?
+
+      Loan::InterestAccrual.new(account.loan).sync!
+    rescue => e
+      Rails.logger.error("Error accruing loan interest for account #{account.id}: #{e.message}")
+      Sentry.capture_exception(e)
```
[the remark this answers is not in the document]
Comment (thewillwilson) [reply app/models/account/syncer.rb:47]: Agreed and fixed in e8bd8f88 — this was a straight miss against the guidance in `CLAUDE.md`/`AGENTS.md`.

`Account::Syncer#accrue_loan_interest` now calls `DebugLogEntry.capture(...)` with `category: "loan_interest_accrual_error"`, `level`, `source`, the `account` and `family` for filtering, and structured metadata, keeping `Sentry.capture_exception` alongside. Since the failure is deliberately non-fatal, the user sees a loan that has quietly stopped accruing with nothing surfaced in the UI, so the support-visible incident matters here.

File: test/models/account/syncer_test.rb [hunk 4]
```diff
@@ -36,4 +36,62 @@ class Account::SyncerTest < ActiveSupport::TestCase
...
   end
+
+  test "accrues loan interest before materializing balances" do
+    account = accounts(:loan)
+
+    Account::MarketDataImporter.any_instance.expects(:import_all).once
+    Loan::InterestAccrual.any_instance.expects(:sync!).once.returns(false)
+    Balance::Materializer.any_instance.expects(:materialize_balances).once
+
+    Account::Syncer.new(account).perform_sync(OpenStruct.new(window_start_date: nil))
```
[the remark this answers is not in the document]
Comment (thewillwilson) [reply test/models/account/syncer_test.rb:51]: Fair — the test asserted both calls happened but never that accrual came first, which is the entire guarantee its name claims. Fixed in e8bd8f88 using a Mocha `sequence`, essentially as suggested.

Worth stating why the ordering is load-bearing rather than cosmetic: accruals have to exist before `Balance::Materializer` runs, or the new entries are not picked up until the *following* sync.

File: app/controllers/api/v1/transactions_controller.rb [hunk 5]
```diff
@@ -14,8 +14,13 @@ def index
       .accessible_by(current_resource_owner)
       .where.not(status: "pending_deletion")
       .select(:id)
+    # System-generated entries (e.g. loan interest accruals) are derived output
+    # that the web surfaces already hide. The API has no `excluded` field for a
+    # client to filter on, so returning them here would silently double-count
+    # against the payments they offset.
     transactions_query = family.transactions
       .joins(:entry).where(entries: { account_id: accessible_account_ids })
+      .merge(Entry.excluding_system_generated)
```
[the remark this answers is not in the document]
Comment (thewillwilson) [reply app/controllers/api/v1/transactions_controller.rb:23]: Confirmed and fixed in 823c92b7 — and the problem was a bit worse than described.

`show` returning a hidden entry is a contract inconsistency rather than a double-count risk on its own (one record, one fetch). But `update` and `destroy` share `set_transaction`, so the gap also meant the API could **mutate** a generated entry — including setting `excluded` back to false. That is precisely the permanent double-count this PR added the web-side containment to prevent, reachable through a second door. `set_transaction` now merges the same scope, so all three actions 404. Test added.

For what it is worth this was a one-line change rather than a heavy lift.

Worth raising for maintainers, since this is the second surface missed for the same reason: `Entry.excluding_system_generated` has to be applied by hand at every entry point, and nothing enforces it. Today that is the web search, rules, both export paths, and now API index + show. A future contributor adding a sixth surface has no signal they need it. If you would rather see a `default_scope`, a distinct entryable type, or the accrual represented some other way entirely, I am glad to rework it — that is a design call I would rather have made explicitly than left as a convention to remember.

File: test/controllers/api/v1/transactions_controller_test.rb [hunk 6]
```diff
@@ -117,6 +117,27 @@ class Api::V1::TransactionsControllerTest < ActionDispatch::IntegrationTest
...
+      external_id: Date.current.strftime("%Y-%m"),
+      entryable: Transaction.new(kind: "standard")
+    )
+
+    get api_v1_transactions_url, headers: api_headers(@api_key)
+    assert_response :success
+
+    transaction_ids = JSON.parse(response.body)["transactions"].map { |t| t["id"] }
+    assert_not_includes transaction_ids, accrual.entryable_id
+  end
```
[the remark this answers is not in the document]
Comment (thewillwilson) [reply test/controllers/api/v1/transactions_controller_test.rb:145]: Fair, and a genuine weakness in the test rather than a nitpick — it used the default 25-item page, so it could have passed because the entry fell off page one rather than because it was filtered.

Fixed in 823c92b7: requests `per_page: 100` and asserts the result set fits within a single page, so the absence actually proves filtering. Also lifted the fixture into a shared helper now that the new `show` 404 test needs it too.

File: app/controllers/api/v1/transactions_controller.rb [hunk 7]
```diff
@@ -204,9 +209,14 @@ def set_transaction
...
 
       family = current_resource_owner.family
+      # Same exclusion as #index, so the contract holds across the whole
+      # resource: a generated entry is a 404 here rather than something a client
+      # can fetch by id — and, since update/destroy share this lookup, not
+      # something the API can mutate (e.g. un-exclude) either.
       @transaction = family.transactions
         .joins(entry: :account)
         .merge(Account.accessible_by(current_resource_owner))
+        .merge(Entry.excluding_system_generated)
```
[the remark this answers is not in the document]
Comment (thewillwilson) [reply app/controllers/api/v1/transactions_controller.rb:233]: Confirmed, and fixed in 1f767d55 — but the suggested fix would have closed only the less harmful half, so I went further.

Filtering `existing_idempotent_entry` stops a client *reading* a generated entry. The bigger problem is the write direction: `idempotency_source` (line 395) takes any caller-supplied string with no allowlist, so a client can **create** an entry with `source: "loan_interest_accrual"`. `Loan::InterestAccrual#reconcile` would then destroy it on the next sync as a stale accrual — the client’s data disappears with no error surfaced.

So reserved namespaces are now refused at the boundary with a 422, alongside the existing “Source requires external_id” check, which closes both directions. The lookup is scoped too, as defence in depth. Three tests: reserved source rejected with no entry created, and the idempotency lookup never resurfacing an accrual.

---

Worth stating plainly for maintainers, because this is now the **third** surface of the same class in this review — API index, then show/update/destroy, now the idempotency lookup. Add the four web-side call sites and that is seven places `Entry.excluding_system_generated` has to be remembered, with nothing enforcing it. Each individual fix is a one-liner, which is exactly what makes the pattern easy to keep missing.

I do not think “apply the scope in one more place” is the right long-term answer, and I would rather not keep presenting it as if it were. The options as I see them: a `default_scope` on `Entry`, a distinct entryable type so generated rows are not `Transaction`s at all, or not materialising entries in the first place and computing the interest inside the balance calculator. All are larger changes than this PR, and which one is right is a maintainer call.

It is also fair to note this is a structural point in favour of #2840, which solves the same user problem without creating system-owned entries — so it has no containment surface to leak from at all.

File: test/controllers/api/v1/transactions_controller_test.rb [hunk 8]
```diff
@@ -117,6 +117,76 @@ class Api::V1::TransactionsControllerTest < ActionDispatch::IntegrationTest
...
+             name: "Idempotent replay",
+             external_id: accrual.external_id
+           }
+         },
+         headers: api_headers(@api_key)
+
+    # Must not echo back the generated entry; it either creates a new
+    # transaction or fails, but never resurfaces the accrual.
+    assert_not_equal accrual.entryable_id, JSON.parse(response.body)["id"]
+  end
```
[the remark this answers is not in the document]
Comment (thewillwilson) [reply test/controllers/api/v1/transactions_controller_test.rb]: You are right on both counts, and the test was worse than described — fixed in ac90c451.

It omitted `source`, so the lookup ran under `"api"` and could never match the accrual’s `"loan_interest_accrual"`. It would have passed with the scoping reverted entirely. And `assert_not_equal` passed trivially against an error response carrying no `id`. It asserted nothing.

There is a further wrinkle worth recording: now that reserved sources are refused at the boundary with a 422, the scoped lookup is **unreachable through the API**. There is no honest end-to-end assertion for it — it is defence in depth for a future caller path only, and I would rather say that than dress up a test that appears to cover it.

Replaced with a test that pins real behaviour instead: a client reusing a generated entry’s `external_id` under its own source is a *distinct* key `(account, "api", "2026-07")` and must still create successfully, so the reserved namespace does not shadow legitimate clients. That one fails if the boundary check over-reaches, which the previous test could not detect.

**Comment by sure-design** [comment]
Thanks for flagging the overlap with #2840 so transparently yourself. Agreed with your read: #2840 was opened first, already has @jjmata's review, and matches what #1804 literally asked for (a real "Loan Interest" category visible in cash flow/budgets), where this PR's design intentionally keeps the accrual out of `IncomeStatement`. Recommend closing this one in favor of #2840.

One caveat worth noting for whoever picks this up: #2840 currently has its own unresolved data-integrity concern flagged by @jjmata (`Loan::PaymentResplitter` destroys the original transfer/entry *before* checking whether the re-split is applicable, which can silently drop a loan's principal reduction with no rollback) — that needs fixing regardless of which design lands.

Separately, the `Entry#linked?` / empty-provider "Linked with " badge fix you called out is independent of the loan-design decision — worth splitting into its own small PR so it isn't held hostage to this one.

Suggested line for AGENTS.md/CLAUDE.md: *"Before starting a new feature branch, search open PRs by problem domain and affected models (not just literal issue-title wording) to catch overlapping in-flight work before writing code."*

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by thewillwilson** [comment]
@jjmata — flagging this for a human steer, because I think an inaccuracy of mine has been feeding back around the loop.

**First, correcting myself.** In my comment above I wrote that #2840 "already has @jjmata's review," in a paragraph arguing this PR should give way. That framing was wrong, and I should have been more careful with it. Your review of #2840 was *blocking*, not approving:

> I'd want the resplitter destructive-ordering issue addressed before this ships, since it can quietly corrupt loan history with no error surfaced to the user.

#2840 is still on its original single commit (`ba577fb`, 2026-07-28T16:03) — nothing has been pushed since that review on 2026-07-29T01:13, so the destroy-before-checking-applicability issue and CodeRabbit's seven findings are all still open. The follow-up here then cited "already has @jjmata's review" back to me as grounds for closing this PR. That's my error propagating, not new information, and I'd rather flag it than let it quietly settle the question.

Also worth being explicit for anyone skimming: every review on both PRs so far is automated. No human has looked at this one.

## Where I've landed on the two designs

I still think #2840 is better on the thing #1804 actually asked for, and I said so above — I'm not walking that back. But having now read both closely, I don't think it's either/or, and the split isn't where I originally put it.

**#2840 gets the representation right, and this PR does not.** Splitting the payment makes principal a transfer and only interest an expense, so interest lands in cash flow and budgets as a real category. This PR deliberately keeps the accrual out of `IncomeStatement` — the payment is already booked as a `loan_payment` expense, so counting interest again would double it — which means interest only ever surfaces on the loan's overview card. A user still can't answer "what did my mortgage cost me this year?" from reports. That's the headline ask, and this PR doesn't deliver it.

**This PR gets the accrual right, and #2840 does not.**

| | #2842 | #2840 |
|---|---|---|
| User's payment transaction | never touched | destroyed and recreated |
| Interest basis | replayed day-by-day from the balance actually outstanding on each accrual date | computed once at match time |
| Variable / adjustable rates | reads the configured rate each period | effectively fixed-rate |
| Provider-synced loans | refuses to accrue at all (`account.unlinked?` gate) | rewrites synced transactions |
| Re-running it | idempotent reconcile; converges regardless of order | destructive resplit |

The destructive-ordering bug you found is fixable with check-then-destroy, and I'd expect that fix to land. But the thing underneath it doesn't go away: rewriting the user's own transactions means every future interaction — bank sync overwriting a split payment, transfer matching, the split machinery, an import replaying history — is a chance to corrupt data that the user entered. Your second observation on that PR is the same root cause showing up again: back-to-back matches computing interest against a stale materialized balance. This PR sidesteps both by never mutating user rows and re-deriving state from scratch each sync.

**Honest cost of this design, stated plainly:** it creates system-owned entries, and containing them means calling `Entry.excluding_system_generated` at seven call sites with nothing enforcing it. Three of those leaks were found by reviewers after I thought the work was done. That's a real structural wart and it's the strongest argument against this approach — I raised it in the PR body and I'll raise it again here rather than let it slide.

## What I'd actually suggest

They're complementary. The accrual replay answers *when and how much*; the payment split answers *how it shows up*. The version I'd want as a user is this PR's engine feeding #2840's representation — accrue against the real outstanding balance so variable rates and irregular payments are correct, then present the payment as principal-transfer-plus-interest-expense so it reaches cash flow.

I'm not asking anyone to arbitrate that on my say-so, and I'm still happy to close this in favour of #2840 — the offer above stands unconditionally. What I'd like is a maintainer's steer on the direction before either of us builds more:

1. Is representation-in-cash-flow the requirement (→ #2840's model wins, and I'd contribute the accrual engine into it), or is not-mutating-user-transactions the requirement (→ this model, and I close the cash-flow gap here)?
2. Is @Tynman88 still on #2840? It hasn't moved since your review. If it's stalled I'd rather help fix the resplitter than have two of us stalled in parallel.

Whichever way it goes, the `Entry#linked?` empty-provider badge fix is unrelated to loans and I'll pull it into its own PR so it isn't held up by this.

CI on this PR is green across the board and it's mergeable, for what that's worth — but I'd rather it be judged on the design question than on being ready.

**Review by jjmata — COMMENTED** [review]
Reviewed the diff, the linked issue (#1804), and CI (all green). This is a well-engineered change — thorough comments, `BigDecimal` throughout (no float rounding risk), `Money` used correctly for the totals, `DebugLogEntry` + `Sentry` on the swallowed sync failure per repo convention, the accrual logic lives as a PORO under `app/models/loan/` (fits "fat models, skinny controllers/no services"), and test coverage in `test/models/loan/interest_accrual_test.rb` is unusually thorough — idempotency, month-end clamping, valuation-date skipping, custom FX rates, payoff, linking/opt-out purges, and a direct regression test for the backdated-anchor scenario described in #1804.

**Main correctness question — rate changes are retroactive across all history, not just from the change date forward.**

`Loan::InterestAccrual#desired_accruals` (`app/models/loan/interest_accrual.rb`) always prices every historical period using `loan.monthly_interest_rate`, i.e. the *currently configured* rate — there's no effective-dated rate log. This is exercised deliberately in `test/models/loan/interest_accrual_test.rb` ("re-prices existing accruals when the interest rate changes"), where a rate bump retroactively repriced a period that already occurred under the old rate.

That's fine for a data-entry correction (typo fix), but `accrues_interest?` and its comments explicitly claim support for variable/adjustable-rate loans ("works for variable and adjustable rates... a variable or adjustable rate simply takes effect from the next accrual onward" — `app/models/loan.rb`). For a genuine ARM reset, this implementation will rewrite the interest amount (and therefore the derived balance) for *every* period since inception using the new rate, not just periods from the reset date forward. That contradicts both the lender's actual statements and the "self-correct on the next accrual" framing in the PR description. Worth deciding explicitly: either document this as a known limitation (fixed-rate only, in practice), or price each period using the rate that was configured at the time (would need a small effective-dated rate history rather than a single `interest_rate` column).

**Minor items:**
- `app/controllers/api/v1/transactions_controller.rb` gained real behavior changes (new `422 "Source is reserved"` branch on create, and `excluding_system_generated` filtering on index/show/update/destroy), but `spec/requests/api/v1/transactions_spec.rb` wasn't touched. The PR body says "No API surface, so no `spec/requests/api/v1/` changes," but per this repo's mandatory API-consistency convention, a behavior change to an existing endpoint (new error case + changed filtering) should get a docs-only rswag entry even without a schema/field change, so `docs/api/openapi.yaml` reflects the new validation error.
- `desired_accruals` replays the account's full ledger day-by-day from `opening_anchor_date` to `Date.current` on *every* sync call (not just when something changed), which is O(days since origination) — for a 30-year mortgage that's ~11k iterations per sync. Scope is limited to opted-in, unlinked loans so frequency should be low, but worth a comment either confirming this was measured as negligible or flagging as a future optimization target if `SyncAllJob` cadence increases.

No blocking issues found in balance materialization, idempotency (external_id keyed by period, checked against the "one accrual per month" invariant), migration safety (additive, opt-in, no backfill needed since default `accrue_interest: false`), or the `excluded`/`IncomeStatement` double-counting concern the PR calls out — all look correctly handled and tested.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by jjmata** [comment]
Let's see what @Tynman88 says, thanks for the thorough analysis!

**Comment by sure-admin** [comment]
Hi @thewillwilson, this PR currently has merge conflicts with the base branch. Please rebase or merge the latest base branch and resolve the conflicts when you get a chance.

As part of the PR backlog cleanup, maintainers plan to close PRs that still need author action in one week. If you want to keep working on this or need more time, just comment here and we will leave it open. Thanks!

**Comment by Tynman88** [comment]
@thewillwilson @jjmata — chiming in as the author of #2840.

Having read this one properly, I think your accrual approach is the better foundation for #1804: it never rewrites the user's payment, re-derives interest from the balance actually outstanding each period, and avoids the destructive-rewrite failure mode mine had. (I just pushed a fix for the specific data-integrity bug @jjmata caught over on #2840 — the resplitter destroying the original transfer before checking whether the re-split applied — but the whole class of that simply doesn't exist here, since you never touch the payment.)

The one place I'd still argue #2840 got it right is *representation*: splitting the payment makes interest a real expense category visible in cash flow and budgets, which is the headline of #1804. This PR intentionally keeps the charge out of `IncomeStatement`, so interest only surfaces on the loan overview card — a user still can't answer "what did my mortgage cost me this year?" from reports. If there's a clean way to surface the accrued interest as a real category without double-counting it against the `loan_payment` expense already booked on the funding account, I think that closes the gap and this becomes the complete answer to the issue.

Either way, I'm not going to keep pushing #2840 as a competing design — happy to see this one move forward. Thanks for flagging the overlap so transparently up front.

**Comment by sure-design** [comment]
Status update: the author of the competing #2840 has now conceded (Aug 11) that this accrual-based approach is the stronger foundation for #1804, resolving the earlier need for a maintainer steer between the two designs. The remaining blocker is the merge conflict with main flagged by the bot on Aug 7 (`mergeable_state: unknown`) — CI was green before that, and jjmata's substantive review concern (retroactive rate repricing across all history) has since been addressed via the effective-dated `rate_changes` table per the PR description. Worth a rebase to confirm mergeable_state and CI, then this looks close to ready.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by sure-design** [comment]
Triage note: this PR and #2984 (native amortization schedules) both add interest-posting logic to the same `Account::Syncer#perform_sync` integration point, and #2985 (fixed-return deposits) does too. Nobody's flagged the three-way overlap in one place yet — worth a maintainer decision on landing order / consolidation before any of them merge, on top of the rebase this PR needs to clear its current conflicts.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by thewillwilson** [comment]
Rebased onto latest `main` and resolved the conflicts (locale + schema). All CI is green (unit, system, lint, security), and the full suite passes locally (6999 runs, 0 failures). The PR shows mergeable again.

The substantive review items are addressed on the branch:
- @jjmata's retroactive-rate-repricing concern → handled by the effective-dated `Loan::RateChange` model (interest before a change keeps the old rate; only months on/after the effective date use the new one).
- The two Codex threads (accrual-settings round-trip on import, and excluding generated entries from `GET /api/v1/transactions`) were fixed in `e8bd8f8` with tests — fine to mark resolved.
- New `422 "Source is reserved"` case documented; `desired_accruals` replay cost commented.

On @Tynman88's point about surfacing interest as a real expense category in cash flow/budgets: I'm intentionally **deferring that to a follow-up** rather than adding it here. Doing it without double-counting the already-booked `loan_payment` expense means splitting the payment (principal→transfer, interest→expense) — i.e. #2840's representation model — which re-introduces the payment-mutation surface this PR deliberately avoids. Cleaner to land this as the accrual engine and let representation get its own review, ideally this engine feeding that split representation.

Also happy to take a maintainer steer on landing order relative to #2984/#2985, which touch the same `perform_sync` integration point.

**Comment by sure-design** [comment]
Great to see the retroactive-repricing concern resolved via the effective-dated `Loan::RateChange` model. Two things: `mergeable_state` shows dirty again despite the Aug 24 rebase, so it likely needs another rebase against current main. Also, this PR, #2840, #2984, and #2985 all touch `Account::Syncer#perform_sync` for loan interest — could we get a landing-order decision across the four so they don't conflict with each other?

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by sure-design** [comment]
Update: a maintainer merge commit (`8fd6aac`, rebasing onto current `main`) landed today. `mergeable_state` is now `blocked` rather than `dirty` — the conflict is resolved — and `ci/*` has actually run and is fully green (lint, test_unit, test_system, scan_ruby, scan_js). The retroactive-rate-repricing concern from the earlier review is addressed via the effective-dated `Loan::RateChange` model.

No code blocker remains here. Still worth a landing-order call relative to #2984 and #2985, which touch the same `Account::Syncer#perform_sync` integration point.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by thewillwilson** [comment]
@jjmata What's the plan for this? We've got 4/5 PRs open for similar functionality
